src/umpire/util/AllocationMap.cpp

```diff
@@ -14,249 +14,481 @@
+         UMPIRE_LOG(Debug, "Found " << ptr << " at " << parent_ptr
+                    << " with size " << alloc_record->m_size);
+      }
+      else {
+         alloc_record = nullptr;
       }
-    } else {
-      UMPIRE_ERROR("Cannot remove " << ptr );
     }
-
     UMPIRE_UNLOCK;
-  } catch (...) {
+  }
+  catch (...){
     UMPIRE_UNLOCK;
     throw;
   }
 
-  return ret;
+  return alloc_record;
 }
 
-AllocationRecord*
-AllocationMap::findRecord(void* ptr) const
+AllocationRecord AllocationMap::remove(void* ptr)
 {
-
-  Entry alloc_record = nullptr;
+  AllocationRecord ret{};
 
   try {
     UMPIRE_LOCK;
-    auto record = m_records->atOrBefore(reinterpret_cast<uintptr_t>(ptr));
-    if (record.value) {
-      void* parent_ptr = reinterpret_cast<void*>(record.key);
-      alloc_record =
-        reinterpret_cast<Entry>(record.value->back());
 
-      if (alloc_record &&
-          ((static_cast<char*>(parent_ptr) + alloc_record->m_size)
-             > static_cast<char*>(ptr))) {
+    UMPIRE_LOG(Debug, "Removing " << ptr);
 
-         UMPIRE_LOG(Debug, "Found " << ptr << " at " << parent_ptr
-            << " with size " << alloc_record->m_size);
+    // Locate ptr
+    m_last = judy_slot(m_array, reinterpret_cast<unsigned char*>(&ptr), m_depth * JUDY_key_size);
 
+    // If found, remove it
+    if (m_last && *m_last) {
+      RecordList* list = reinterpret_cast<RecordList*>(*m_last);
+      UMPIRE_ASSERT(list->size());
+
+      ret = list->pop_back();
+
+      --m_size;
+
+      if (list->empty()) {
+        if((m_last = judy_slot(m_array, reinterpret_cast<unsigned char*>(&ptr), m_depth * JUDY_key_size)) != nullptr) {
+          auto list = reinterpret_cast<RecordList*>(*m_last);
+
+          // Manually call destructor
+          list->~RecordList();
+
+          // Mark as deallocated in the pool
+          list_pool.deallocate(list);
+
+          m_last = judy_del(m_array);
+        }
       }
-      else {
-         alloc_record = nullptr;
-      }
+    } else {
+      UMPIRE_ERROR("Cannot remove " << ptr );
     }
+
     UMPIRE_UNLOCK;
-  }
-  catch (...){
+  } catch (...) {
     UMPIRE_UNLOCK;
     throw;
   }
 
-  return alloc_record;
+  return ret;
 }
 
-AllocationRecord*
-AllocationMap::find(void* ptr) const
+bool AllocationMap::contains(void* ptr) const
 {
   UMPIRE_LOG(Debug, "Searching for " << ptr);
-
-  auto alloc_record = findRecord(ptr);
-
-  if (alloc_record) {
-    return alloc_record;
-  } else {
-#if !defined(NDEBUG)
-    // use this from a debugger to dump the contents of the AllocationMap
-    printAll();
-#endif
-    UMPIRE_ERROR("Allocation not mapped: " << ptr);
-  }
-}
-
-AllocationMap::ConstIterator AllocationMap::begin() const {
-  return ConstIterator(m_records, false);
+  return (findRecord(ptr) != nullptr);
 }
 
-AllocationMap::ConstIterator AllocationMap::end() const {
-  return ConstIterator(m_records, true);
-}
+void AllocationMap::clear()
+{
+  uintptr_t key = 0;
+  // TODO Why is max = 0 here?
```

Comment:
Either figure this out and update the comment, or remove the comment and create a JIRA ticket.